llmfoundry/models/inference_api_wrapper/openai_causal_lm.py

```diff
@@ -0,0 +1,239 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Implements a Hugging Causal LM wrapped inside a :class:`.ComposerModel`."""
+
+import os
+from time import sleep
+from typing import Any, Dict, List, Optional, Union
+
+import openai
+import tiktoken
+import torch
+from composer.core.types import Batch
+from openai.error import RateLimitError
+from transformers import PreTrainedTokenizer, PreTrainedTokenizerFast
+
+from llmfoundry.models.inference_api_wrapper.interface import \
+    InferenceAPIEvalWrapper
+
+__all__ = [
+    'OpenAICausalLMEvalWrapper', 'OpenAIChatAPIEvalWrapper',
+    'OpenAITokenizerWrapper'
+]
+
+Tokenizer = Union[PreTrainedTokenizer, PreTrainedTokenizerFast]
+openai.api_key = os.getenv('OPENAI_API_KEY')
+
+MAX_RETRIES = 100
+
+
+class OpenAITokenizerWrapper:
+
+    def __init__(self, name: str) -> None:
+        self.tokenizer = tiktoken.encoding_for_model(name)
+
+    def __call__(self, x: str, add_special_tokens: bool = False):
+        return self.encode(x)
+
+    def encode(self,
+               x: Union[str, List[str]],
+               add_special_tokens: bool = False):
+        if isinstance(x, str):
+            return {
+                'input_ids':
+                    self.tokenizer.encode(x, allowed_special={'<|endoftext|>'})
+            }
+        elif isinstance(x,
+                        list):  # pyright: ignore [reportUnnecessaryIsInstance]
+            return {
+                'input_ids':
+                    self.tokenizer.encode_batch(
+                        x, allowed_special={'<|endoftext|>'})
+            }
+        else:
+            raise ValueError(
+                f'`encode` argument must be str or List[str], got: {type(x)}')
+
+    def decode(self, x: List[int]):
```

Comment:
should this accept `List[int]` or `List[List[int]]` to match the the `encode` logic?